### PR TITLE
Clean library KODI after PP finishes

### PR DESCRIPTION
**Pull request by fernandog**
- [x] TODO: adding UI option (default off)

Usefull when we have a redownload (after failed or after manual search or when higher quality found) and Kodi shows two episodes in the library but only one really exists because the other was replaced in medusa

@ratoaq2 i know you use kodi, do what do you think? the idea is to add an option to this

i tested and its working. Last PP log is: "Successfully processed"

```
2016-08-15 00:10:49 DEBUG    Thread-17 :: [f3ebd83] KODI JSON response: {u'jsonrpc': u'2.0', u'id': 1, u'result': u'OK'}
2016-08-15 00:10:42 DEBUG    Thread-17 :: [f3ebd83] Contacting KODI (with auth header) via url: http://localhost:0000/jsonrpc
2016-08-15 00:10:42 DEBUG    Thread-17 :: [f3ebd83] KODI JSON command: {"jsonrpc":"2.0","method":"VideoLibrary.Clean","id":1}
2016-08-15 00:10:42 DEBUG    Thread-17 :: [f3ebd83] Updating KODI library via JSON method for host: localhost:0000
2016-08-15 00:10:42 INFO     Thread-17 :: [f3ebd83] Successfully processed

```

### CODE REVIEW

**Comment by ratoaq2** [comment]
Library clean might take long time and consume resources. I have it schedule to run every day in the morning. You should at least have a way to turn it off if you don't want a library clean after every single download. And it will be better if it happens only when it's replacing an episode with another one.

**Comment by medariox** [comment]
I think this is a nice feature. I use KODI myself and I know that this is a real issue. Anyhow, @ratoaq2 is right, cleaning the library is a resource intensive process that can take some time, especially on big libraries and low spec devices. I would keep the feature disabled by default, with the option to enable it.

**Comment by ratoaq2** [comment]
Yeah, it's a nice feature 

Probably the options should be: 
disabled,  every download, every down replacement (if it's possible or easy to distinguish them)

File: sickbeard/notifiers/kodi.py [hunk 1]
```diff
@@ -404,6 +404,22 @@ def _send_to_kodi_json(command, host=None, username=None, password=None, dest_ap
                 logger.log(u"Warning: Couldn't contact %s JSON API at %s: %r" % (dest_app, ss(url), ex(e)), logger.WARNING)
             return False
 
+    def clean_library(self):
+        """Handles clean library KODI host via HTTP JSON-RPC."""
+        for host in [x.strip() for x in sickbeard.KODI_HOST.split(",")]:
+            logger.log(u"Updating KODI library via JSON method for host: {0}".format(host), logger.DEBUG)
+            updateCommand = '{"jsonrpc":"2.0","method":"VideoLibrary.Clean","id":1}'
```
Comment (ratoaq2) [inline sickbeard/notifiers/kodi.py]: Snake case: 🐍

**Comment by fernandog** [comment]
Dont know if we can distinguish if its a replace or not.

I dont have a huge library but I use a rpi. My library clean takes like 5s.

**Comment by fernandog** [comment]
@medariox @ratoaq2 check new commit. I dont see another solution for that

File: sickbeard/postProcessor.py [hunk 2]
```diff
@@ -1061,7 +1061,7 @@ def process(self):  # pylint: disable=too-many-return-statements, too-many-local
         for cur_ep in [ep_obj] + ep_obj.related_episodes:
             try:
                 self._delete(cur_ep.location, associated_files=True)
-
+                sickbeard.KODI_LIBRARY_CLEAN_PENDING = True
```
Comment (ratoaq2) [inline sickbeard/postProcessor.py]: Why we have this here?

Comment (fernandog) [inline sickbeard/postProcessor.py]: this is the part of code where we delete the replaced file

Comment (ratoaq2) [inline sickbeard/postProcessor.py]: But it's always true and it doesn't seem to be the place to change a config value

Comment (fernandog) [inline sickbeard/postProcessor.py]: its a global var. not config var. In __init__ it start as True.

So when we call the self_delete it delete the old file then I set to clean library

Comment (ratoaq2) [inline sickbeard/postProcessor.py]: And how to disable it?

File: gui/slick/views/config_notifications.mako [hunk 3]
```diff
@@ -94,11 +94,20 @@
                                     </label>
                                 </div>
                                 <div class="field-pair">
+                                    <label for="kodi_clean_library">
+                                        <span class="component-title">Clean library</span>
+                                        <span class="component-desc">
+                                            <input type="checkbox" name="kodi_clean_library" id="kodi_clean_library" ${'checked="checked"' if sickbeard.KODI_CLEAN_LIBRARY else ''}/>
+                                            <p>clean KODI library when replaces a already download episode?</p>
```
Comment (ratoaq2) [inline gui/slick/views/config_notifications.mako]: minor type: should be `already downloaded`

File: sickbeard/notifiers/kodi.py [hunk 4]
```diff
@@ -404,6 +404,28 @@ def _send_to_kodi_json(command, host=None, username=None, password=None, dest_ap
                 logger.log(u"Warning: Couldn't contact %s JSON API at %s: %r" % (dest_app, ss(url), ex(e)), logger.WARNING)
             return False
 
+    def clean_library(self):
+        """Handles clean library KODI host via HTTP JSON-RPC."""
+        clean_library = True
+        for host in [x.strip() for x in sickbeard.KODI_HOST.split(",")]:
```
Comment (ratoaq2) [inline sickbeard/notifiers/kodi.py]: Change double quotes to single quotes in this method.

File: sickbeard/notifiers/kodi.py [hunk 5]
```diff
@@ -404,6 +404,28 @@ def _send_to_kodi_json(command, host=None, username=None, password=None, dest_ap
                 logger.log(u"Warning: Couldn't contact %s JSON API at %s: %r" % (dest_app, ss(url), ex(e)), logger.WARNING)
             return False
 
+    def clean_library(self):
+        """Handles clean library KODI host via HTTP JSON-RPC."""
+        clean_library = True
+        for host in [x.strip() for x in sickbeard.KODI_HOST.split(',')]:
+            logger.log(u"Updating KODI library via JSON method for host: {0}".format(host), logger.DEBUG)
```
Comment (ratoaq2) [inline sickbeard/notifiers/kodi.py]: log messages are still double quotes. The rest looks fine

**Comment by fernandog** [comment]
@medariox @ratoaq2 I manually download a torrent and manually PP it again (force=yes)

it worked:

```
2016-08-16 16:31:45 DEBUG    Thread-16 :: [f3ebd83] KODI JSON response: {u'jsonrpc': u'2.0', u'id': 1, u'result': u'OK'}
2016-08-16 16:31:37 DEBUG    Thread-16 :: [f3ebd83] Contacting KODI (with auth header) via url: http://localhost:0000/jsonrpc
2016-08-16 16:31:37 DEBUG    Thread-16 :: [f3ebd83] KODI JSON command: {"jsonrpc":"2.0","method":"VideoLibrary.Clean","id":1}
2016-08-16 16:31:37 DEBUG    Thread-16 :: [f3ebd83] Updating KODI library via JSON method for host: localhost:0000
2016-08-16 16:31:37 INFO     Thread-16 :: [f3ebd83] Successfully processed
2016-08-16 16:31:37 INFO     Thread-16 :: [f3ebd83] Not deleting folder /media/SAMSUNG/media/downloaded/series/*.S04.1080p.AMZN.WEBRip.DD5.1.x264-NTb found the following files: [u'*.1080p.AMZN.WEBRip.DD5.1.x264-NTb.mkv']
2016-08-16 16:31:37 INFO     Thread-16 :: [f3ebd83] Processing succeeded for /media/SAMSUNG/media/downloaded/series/*.S04.1080p.AMZN.WEBRip.DD5.1.x264-NTb/*.1080p.AMZN.WEBRip.DD5.1.x264-NTb.mkv
2016-08-16 16:31:37 INFO     Thread-16 :: [f3ebd83] File exists and new file is same size, pretending we did something
```

**Comment by ratoaq2** [comment]
Now it works:
```
2016-08-17 21:24:58 DEBUG    Thread-17 :: [58bd644] Cleaning KODI library via JSON method for host: kodi-server:80
2016-08-17 21:24:58 DEBUG    Thread-17 :: [58bd644] KODI JSON command: {"jsonrpc":"2.0","method":"VideoLibrary.Clean","params": {"showdialogs": false},"id":1}
2016-08-17 21:24:58 DEBUG    Thread-17 :: [58bd644] Contacting KODI via url: http://kodi-server:80/jsonrpc
2016-08-17 21:24:58 DEBUG    Thread-17 :: [58bd644] KODI JSON response: {u'jsonrpc': u'2.0', u'id': 1, u'result': u'OK'}
```

```
21:25:14 1146241.500000 T:793768944  NOTICE: CleanDatabase: Starting videodatabase cleanup ..
```

**Comment by ratoaq2** [comment]
Approved

File: sickbeard/postProcessor.py (hunk 2 above)
Comment (fernandog) [inline sickbeard/postProcessor.py]: @medariox here. We set the clean var to true after we deleted a file

**Pull request merged**